File: generators/reference/kernels/ffn_dense.py
"""ffn.dense@1.0.0 — act(x·inᵀ + in_bias) · outᵀ + out_bias: the up projection, the activation,
the down projection — `ffn.gated` without its gate half.

| branch / record           | status                 |
|---------------------------|------------------------|
| activation silu           | implemented            |
| activation gelu           | implemented (erf)      |
| activation gelu_tanh      | implemented            |
| activation relu2          | implemented            |
| in_bias, out_bias         | implemented            |
| condition_dim / condition | refused                |
| activation_sparsity       | refused when > 0       |
"""
import torch.nn.functional as F
from kernels._common import refuse_unknown, supports_from, w

CONTRACT = ("ffn.dense", "1.0.0")
ACT = {'silu': F.silu, 'gelu': F.gelu, 'gelu_tanh': lambda x: F.gelu(x, approximate='tanh'),
       'relu2': lambda x: F.relu(x).pow(2)}


CAPABILITIES = {"arguments": {"width": "any", "inner": "any", "activation": ["silu", "gelu", "gelu_tanh", "relu2"],
                              "in_bias": "any", "out_bias": "any", "condition_dim": "absent", "condition": "absent",
                              "activation_sparsity": {"absent": True, "values": [0, 0.0]}},
                "states": []}


# What a conformer must meet against this kernel's unit fixtures, per compute dtype (§4.2):
# `|a − b| ≤ atol + rtol·|b|`. The manifest's witness block is written from it.
TOLERANCE = {'f32': {'atol': 1e-5, 'rtol': 1e-4}, 'bf16': {'atol': 1e-1, 'rtol': 2e-2}}

# The unit fixtures this kernel produces (docs/TENSORSPINE-FIXTURE.md): one case per branch
# worth its own evidence, at small quantities.
FIXTURES = [
    {"case": "gelu-biased", "seed": 31, "invocations": [{"input": 5}, {"input": 3}],
     "arguments": {"width": 64, "inner": 128, "activation": "gelu", "in_bias": True, "out_bias": True}},
    {"case": "silu", "seed": 32, "invocations": [{"input": 5}], "arguments": {"width": 64, "inner": 128, "activation": "silu"}},
]


def supports(arguments):
    return supports_from(CAPABILITIES, arguments)

def run(ctx, arguments, inputs, params, states, physical=None):
    x = inputs['input']
    h = x @ w(ctx, params['in']).T
    if arguments.get('in_bias'):
        h = h + w(ctx, params['in_bias'])
    y = ACT[arguments['activation']](h) @ w(ctx, params['out']).T
    if arguments.get('out_bias'):
        y = y + w(ctx, params['out_bias'])
    return {'output': y}

File: generators/zml/primitives/ffn_dense.zig
//! ffn.dense@1.0.0 — act(x·inᵀ + in_bias) · outᵀ + out_bias.
//!
//! `ffn.gated` without its gate half: the up projection, the activation, the down
//! projection.
//!
//! | branch                   | status                        |
//! |--------------------------|-------------------------------|
//! | activation silu          | implemented                   |
//! | activation gelu (erf)    | implemented — see activation.zig |
//! | activation gelu_tanh     | implemented                   |
//! | activation relu2         | implemented                   |
//! | in_bias, out_bias        | implemented                   |
//! | activation_sparsity      | refused when > 0              |

const activation = @import("../activation.zig");
const p = @import("../primitive.zig");

pub const primitive: p.Primitive = .{
    .name = "ffn.dense",
    .version = "1.0.0",
    .run = run,
    .capabilities =
    \\{"arguments": {"width": "any", "inner": "any",
    \\               "activation": ["silu", "gelu", "gelu_tanh", "relu2"],
    \\               "in_bias": "any", "out_bias": "any",
    \\               "activation_sparsity": {"absent": true, "values": [0, 0.0]}},
    \\ "states": [],
    \\ "notes": ["activation gelu computes erf by Abramowitz & Stegun 7.1.26 (|e| <= 1.5e-7): neither StableHLO nor ZML binds an erf op"]}
    ,
};

fn run(ctx: *p.Ctx, call: p.Call) ![]const p.Binding {
    const x = p.features(call.inputs.must("input"));
    const name = call.argStr("activation") orelse return p.Error.MissingArgument;

    var h = p.linear(x, call.params.must("in"));
    if (call.argBool("in_bias")) {
        h = h.add(p.features(call.params.must("in_bias")).convert(h.dtype()).broad(h.shape()));
    }

    var y = p.linear(activation.apply(name, h) catch return p.Error.Unimplemented, call.params.must("out"));
    if (call.argBool("out_bias")) {
        y = y.add(p.features(call.params.must("out_bias")).convert(y.dtype()).broad(y.shape()));
    }
    return p.one(ctx, "output", y);
}
